File: app/src/main/java/com/netease/nis/alivedetecteddemo/MainActivity.java
package com.netease.nis.alivedetecteddemo;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.AssetFileDescriptor;
import android.graphics.PixelFormat;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.ViewStub;
import android.view.WindowManager;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.github.mmin18.widget.RealtimeBlurView;
import com.netease.nis.alivedetected.ActionType;
import com.netease.nis.alivedetected.AliveDetector;
import com.netease.nis.alivedetected.DetectedListener;
import com.netease.nis.alivedetected.NISCameraPreview;
import com.sfyc.ctpv.CountTimeProgressView;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import pl.droidsonroids.gif.GifImageView;

import static com.netease.nis.alivedetected.ActionType.ACTION_ERROR;


public class MainActivity extends Activity {
    private final String TAG = "AliveDetector";
    private ViewStub vsStep2, vsStep3, vsStep4;
    private TextView tvStateTip, tvErrorTip, tvStep1, tvStep2, tvStep3, tvStep4;
    private ImageView ivVoice;
    private GifImageView givAction;
    private CountTimeProgressView mCountTimeView;
    private NISCameraPreview mCameraPreview;
    private RelativeLayout viewTipBackground;
    private RealtimeBlurView blurView;
    private AliveDetector mAliveDetector;
    private boolean DEBUG = false;
    private boolean isUsedCustomStateTip = true; // 是否使用自定义活体状态文案
    private static String BUSINESS_ID;
    private Map<String, String> stateTipMap = new HashMap();
    private static final String KEY_STRAIGHT_AHEAD = "straight_ahead";
    private static final String KEY_OPEN_MOUTH = "open_mouth";
    private static final String KEY_TURN_HEAD_TO_LEFT = "turn_head_to_left";
    private static final String KEY_TURN_HEAD_TO_RIGHT = "turn_head_to_right";
    private static final String KEY_BLINK_EYES = "blink_eyes";
    private static final String FRONT_ERROR_TIP = "请移动人脸到摄像头视野中间";
    private static final String TIP_STRAIGHT_AHEAD = "请正对手机屏幕\n" +
            "将面部移入框内";//"请将面部移入框内并保持不动";
    private static final String TIP_OPEN_MOUTH = "张张嘴";
    private static final String TIP_TURN_HEAD_TO_LEFT = "慢慢左转头";
    private static final String TIP_TURN_HEAD_TO_RIGHT = "慢慢右转头";
    private static final String TIP_BLINK_EYES = "眨眨眼";
    private int mCurrentCheckStepIndex = 0;
    private ActionType mCurrentActionType = ActionType.ACTION_STRAIGHT_AHEAD;
    private ActionType[] mActions;
    private boolean isOpenVoice = true;
    private MediaPlayer mPlayer = new MediaPlayer();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initView();
        Util.setWindowBrightness(this, WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_FULL);
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (mPlayer.isPlaying()) {
            mPlayer.stop();
        }
        if (mPlayer != null) {
            mPlayer.reset();
            mPlayer.release();
            mPlayer = null;
        }
    }

    @Override
    public void onDestroy() {
        Util.setWindowBrightness(this, WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE);
        if (isFinishing()) {
            if (mAliveDetector != null) {
                mAliveDetector.stopDetect();
                mAliveDetector.destroy();
            }
            if (mCountTimeView != null) {
                mCountTimeView.cancelCountTimeAnimation();
            }
        }
        super.onDestroy();
    }

    private void initView() {
        mCameraPreview = (NISCameraPreview) findViewById(R.id.surface_view);
        mCameraPreview.getHolder().setFormat(PixelFormat.TRANSLUCENT);
        tvStateTip = findViewById(R.id.tv_tip);
        tvErrorTip = findViewById(R.id.tv_error_tip);
        tvStep1 = findViewById(R.id.tv_step_1);
        vsStep2 = findViewById(R.id.vs_step_2);
        vsStep3 = findViewById(R.id.vs_step_3);
        vsStep4 = findViewById(R.id.vs_step_4);
        givAction = findViewById(R.id.gif_action);
        ImageButton imgBtnBack = findViewById(R.id.img_btn_back);
        imgBtnBack.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mAliveDetector != null) {
                    mAliveDetector.stopDetect();
                }
                finish();
            }
        });
        ivVoice = findViewById(R.id.iv_voice);
        ivVoice.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                isOpenVoice = !isOpenVoice;
                if (isOpenVoice) {
                    ivVoice.setImageResource(R.mipmap.ico_voice_open_2x);
                } else {
                    ivVoice.setImageResource(R.mipmap.ico_voice_close_2x);
                }
            }
        });
        mCountTimeView = findViewById(R.id.pv_count_time);
        viewTipBackground = findViewById(R.id.view_tip_background);
        blurView = findViewById(R.id.blur_view);
        initData();
    }

    private void initData() {
        stateTipMap.put(KEY_STRAIGHT_AHEAD, TIP_STRAIGHT_AHEAD);
        stateTipMap.put(KEY_TURN_HEAD_TO_LEFT, TIP_TURN_HEAD_TO_LEFT);
        stateTipMap.put(KEY_TURN_HEAD_TO_RIGHT, TIP_TURN_HEAD_TO_RIGHT);
        stateTipMap.put(KEY_OPEN_MOUTH, TIP_OPEN_MOUTH);
        stateTipMap.put(KEY_BLINK_EYES, TIP_BLINK_EYES);
        BUSINESS_ID = "请输入您的业务id";
        mAliveDetector = AliveDetector.getInstance();
        mAliveDetector.setDebugMode(true);
        mAliveDetector.init(this, mCameraPreview, BUSINESS_ID);
        mAliveDetector.setDetectedListener(new DetectedListener() {
            @Override
            public void onReady(boolean isInitSuccess) {
                if (isInitSuccess) {
                    Log.d(TAG, "活体检测引擎初始化完成");
                } else {
                    //  mAliveDetector.startDetect();
                    Log.e(TAG, "活体检测引擎初始化失败");
                }
            }

            /**
             * 此次活体检测下发的待检测动作指令序列
             *
             * @param actionTypes
             */
            @Override
            public void onActionCommands(ActionType[] actionTypes) {
                mActions = actionTypes;
                String commands = buildActionCommand(actionTypes);
                showIndicatorOnUiThread(commands.length() - 1);
//                showToast("活体检测动作序列为:" + commands);
                Log.e(TAG, "活体检测动作序列为:" + commands);
            }

            @Override
            public void onStateTipChanged(ActionType actionType, String stateTip) {
                Log.d(TAG, "actionType:" + actionType.getActionTip() + " stateTip:" + actionType + " CurrentCheckStepIndex:" + mCurrentCheckStepIndex);
                if (actionType == ActionType.ACTION_PASSED && actionType.getActionID() != mCurrentActionType.getActionID()) {
                    mCurrentCheckStepIndex++;
                    if (mCurrentCheckStepIndex < mActions.length) {
                        updateIndicatorOnUiThread(mCurrentCheckStepIndex);
                        if (isOpenVoice) {
                            playSounds(mCurrentCheckStepIndex);
                        }
                        mCurrentActionType = mActions[mCurrentCheckStepIndex];
                    }
                }
                if (isUsedCustomStateTip) {
                    switch (actionType) {
                        case ACTION_STRAIGHT_AHEAD:
                            setTipText("", false);
                            break;
                        case ACTION_OPEN_MOUTH:
                            setTipText(stateTipMap.get(KEY_OPEN_MOUTH), false);
                            break;
                        case ACTION_TURN_HEAD_TO_LEFT:
                            setTipText(stateTipMap.get(KEY_TURN_HEAD_TO_LEFT), false);
                            break;
                        case ACTION_TURN_HEAD_TO_RIGHT:
                            setTipText(stateTipMap.get(KEY_TURN_HEAD_TO_RIGHT), false);
                            break;
                        case ACTION_BLINK_EYES:
                            setTipText(stateTipMap.get(KEY_BLINK_EYES), false);
                            break;
                        case ACTION_ERROR:
                            setTipText(stateTip, true);
                            break;
//                        case ACTION_PASSED:
//                            setTipText(stateTip);
//                            break;
                    }
                } else {
                    setTipText(stateTip, actionType == ACTION_ERROR);
                }
            }

            @Override
            public void onPassed(boolean isPassed, String token) {
                if (isPassed) {
                    Log.d(TAG, "活体检测通过,token is:" + token);
                    finish();
                    Intent intent = new Intent(MainActivity.this, SuccessActivity.class);
                    startActivity(intent);
                } else {
                    Log.e(TAG, "活体检测不通过,token is:" + token);
                    jump2FailureActivity(token);
                }
            }

            @Override
            public void onError(int code, String msg, String token) {
                Log.e(TAG, "listener [onError] 活体检测出错,原因:" + msg + " token:" + token);
                jump2FailureActivity(token);
            }

            @Override
            public void onOverTime() {
                Util.showDialog(MainActivity.this, "检测超时", "请在规定时间内完成动作",
                        "重试", "返回首页", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                resetIndicator();
                                resetGif();
                                mCountTimeView.startCountTimeAnimation();
                                mAliveDetector.startDetect();
                            }
                        }, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                Intent intent = new Intent(MainActivity.this, WelcomeActivity.class);
                                startActivity(intent);
                            }
                        });
            }
        });
        mAliveDetector.setSensitivity(AliveDetector.SENSITIVITY_NORMAL);
        mAliveDetector.setTimeOut(1000 * 30);
        mAliveDetector.startDetect();
        initCountTimeView();
    }

    private void initCountTimeView() {
        mCountTimeView.setStartAngle(0);
        mCountTimeView.startCountTimeAnimation();
    }

    private void setTipText(final String tip, final boolean isErrorType) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (isErrorType) {
                    if (FRONT_ERROR_TIP.equals(tip)) {
                        tvErrorTip.setText(TIP_STRAIGHT_AHEAD);
                    } else {
                        tvErrorTip.setText(tip);
                    }
                    viewTipBackground.setVisibility(View.VISIBLE);
                    blurView.setVisibility(View.VISIBLE);

                } else {
                    viewTipBackground.setVisibility(View.INVISIBLE);
                    blurView.setVisibility(View.INVISIBLE);
                    tvStateTip.setText(tip);
                    tvErrorTip.setText("");
                }
            }
        });
    }


    public static String buildActionCommand(ActionType[] actionCommands) {
        StringBuilder commands = new StringBuilder();
        for (ActionType actionType : actionCommands) {
            commands.append(actionType.getActionID());
        }
        return TextUtils.isEmpty(commands.toString()) ? "" : commands.toString();
    }

    private void showIndicatorOnUiThread(final int commandLength) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                showViewStub(commandLength);
                showIndicator(commandLength);
            }
        });
    }

    private void showIndicator(int commandLength) {
        switch (commandLength) {
            case 2:
                vsStep2.setVisibility(View.VISIBLE);
                tvStep2.setVisibility(View.VISIBLE);
                break;
            case 3:
                tvStep2.setVisibility(View.VISIBLE);
                tvStep3.setVisibility(View.VISIBLE);
                break;
            case 4:
                tvStep2.setVisibility(View.VISIBLE);
                tvStep3.setVisibility(View.VISIBLE);
                tvStep4.setVisibility(View.VISIBLE);
                break;
        }
    }

    private void showViewStub(int commandLength) {
        switch (commandLength) {
            case 2:
                vsStep2.setVisibility(View.VISIBLE);
                tvStep2 = findViewById(R.id.tv_step_2);
                break;
            case 3:
                vsStep2.setVisibility(View.VISIBLE);
                tvStep2 = findViewById(R.id.tv_step_2);
                vsStep3.setVisibility(View.VISIBLE);
                tvStep3 = findViewById(R.id.tv_step_3);
                break;
            case 4:
                vsStep2.setVisibility(View.VISIBLE);
                tvStep2 = findViewById(R.id.tv_step_2);
                vsStep3.setVisibility(View.VISIBLE);
                tvStep3 = findViewById(R.id.tv_step_3);
                vsStep4.setVisibility(View.VISIBLE);
                tvStep4 = findViewById(R.id.tv_step_4);
                break;
        }
    }

    private void updateIndicatorOnUiThread(final int currentActionIndex) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                updateIndicator(currentActionIndex);
                updateGif(currentActionIndex);
            }
        });
    }

    private void resetIndicator() {
        mCurrentCheckStepIndex = 0;
        mCurrentActionType = ActionType.ACTION_STRAIGHT_AHEAD;
        tvStep1.setText("1");
        if (tvStep2 != null && tvStep2.getVisibility() == View.VISIBLE) {
            tvStep2.setText("");
            setTextViewUnFocus(tvStep2);
        }
        if (tvStep3 != null && tvStep3.getVisibility() == View.VISIBLE) {
            tvStep3.setText("");
            setTextViewUnFocus(tvStep3);
        }
        if (tvStep4 != null && tvStep4.getVisibility() == View.VISIBLE) {
            tvStep4.setText("");
            setTextViewUnFocus(tvStep4);
        }
    }

    private void resetGif() {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                givAction.setImageResource(R.mipmap.pic_front_2x);
            }
        });
    }

    private void updateIndicator(int currentActionPassedCount) {
        switch (currentActionPassedCount) {
            case 2:
                tvStep1.setText("");
                tvStep2.setText("2");
                setTextViewFocus(tvStep2);
                break;
            case 3:
                tvStep1.setText("");
                tvStep2.setText("");
                setTextViewFocus(tvStep2);
                tvStep3.setText("3");
                setTextViewFocus(tvStep3);
                break;
            case 4:
                tvStep1.setText("");
                tvStep2.setText("");
                setTextViewFocus(tvStep2);
                tvStep3.setText("");
                setTextViewFocus(tvStep3);
                tvStep4.setText("4");
                setTextViewFocus(tvStep4);
                break;
        }
    }

    private void updateGif(int currentActionIndex) {
        switch (mActions[currentActionIndex]) {
            case ACTION_TURN_HEAD_TO_LEFT:
                givAction.setImageResource(R.drawable.turn_left);
                break;
            case ACTION_TURN_HEAD_TO_RIGHT:
                givAction.setImageResource(R.drawable.turn_right);
                break;
            case ACTION_OPEN_MOUTH:
                givAction.setImageResource(R.drawable.open_mouth);
                break;
            case ACTION_BLINK_EYES:
                givAction.setImageResource(R.drawable.open_eyes);
                break;
        }
    }

    private void playSounds(int currentActionIndex) {
        switch (mActions[currentActionIndex]) {
            case ACTION_TURN_HEAD_TO_LEFT:
                playSound(getAssetFileDescriptor("turn_head_to_left.wav"));
                break;
            case ACTION_TURN_HEAD_TO_RIGHT:
                playSound(getAssetFileDescriptor("turn_head_to_right.wav"));
                break;
            case ACTION_OPEN_MOUTH:
                playSound(getAssetFileDescriptor("open_mouth.wav"));
                break;
            case ACTION_BLINK_EYES:
                playSound(getAssetFileDescriptor("blink_eyes.wav"));
                break;
        }
    }

    private void setTextViewFocus(TextView tv) {
        tv.setBackgroundDrawable(ContextCompat.getDrawable(getApplication(), R.drawable.circle_tv_focus));
    }

    private void setTextViewUnFocus(TextView tv) {
        tv.setBackgroundDrawable(ContextCompat.getDrawable(getApplication(), R.drawable.circle_tv_un_focus));
    }

    private AssetFileDescriptor getAssetFileDescriptor(String assetName) {
        try {
            AssetFileDescriptor fileDescriptor = getApplication().getAssets().openFd(assetName);
            return fileDescriptor;
        } catch (IOException e) {
            e.printStackTrace();
            Log.e(TAG, "getAssetFileDescriptor error" + e.toString());
        }
        return null;
    }

    private void playSound(AssetFileDescriptor fileDescriptor) {
        try {
            mPlayer.reset();
            mPlayer.setDataSource(fileDescriptor.getFileDescriptor(), fileDescriptor.getStartOffset(), fileDescriptor.getLength());
            mPlayer.prepare();
            mPlayer.start();
        } catch (IOException e) {
            e.printStackTrace();
            Log.e(TAG, "playSound error" + e.toString());
        }
    }

    private void jump2FailureActivity(String token) {
        finish();
        Intent intent = new Intent(MainActivity.this, FailureActivity.class);
        intent.putExtra("token", token);
        startActivity(intent);
    }
}

File: app/src/main/java/com/netease/nis/alivedetecteddemo/MainActivity.kt
package com.netease.nis.alivedetecteddemo

import android.annotation.SuppressLint
import android.content.Intent
import android.content.IntentFilter
import android.content.res.AssetFileDescriptor
import android.media.MediaPlayer
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.netease.cloud.nos.yidun.receiver.ConnectionChangeReceiver
import com.netease.nis.alivedetected.ActionType
import com.netease.nis.alivedetected.AliveDetector
import com.netease.nis.alivedetected.DetectedListener
import com.netease.nis.alivedetected.NISCameraPreview
import com.netease.nis.alivedetecteddemo.manager.BroadcastDispatcher
import com.netease.nis.alivedetecteddemo.utils.Util
import com.netease.nis.alivedetecteddemo.view.FaceAuraColorView
import com.netease.nis.alivedetecteddemo.view.FaceDetectRoundView
import com.netease.nis.alivedetecteddemo.view.SpaceLivenessView
import com.sfyc.ctpv.CountTimeProgressView
import java.io.IOException
import java.util.*

/**
 * @author liu
 * @date 2021/10/12
 * @desc
 * @email liulingfeng@mistong.com
 */
class MainActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "MainActivity"
    }

    private var mAliveDetector: AliveDetector? = null
    private var mActions: Array<ActionType>? = null
    private var mCurrentActionType = ActionType.ACTION_STRAIGHT_AHEAD
    private var isOpenVoice = true
    private var mPlayer: MediaPlayer? = null
    private var progressDialog: AlertDialog? = null
    private var connectionChangeReceiver: ConnectionChangeReceiver? = null

    private var mSpaceLivenessView: SpaceLivenessView? = null
    private var mFaceDetectView: FaceDetectRoundView? = null
    private var imgBtnBack: ImageView? = null
    private var ivVoice: ImageView? = null
    private var surfaceView: NISCameraPreview? = null
    private var pvCountTime: CountTimeProgressView? = null
    private var gifAction: ImageView? = null
    private var detectAura: FaceAuraColorView? = null
    private var tvTip: TextView? = null
    private var mLastSpaceCode = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Util.setWindowBrightness(this, WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_FULL)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        setContentView(R.layout.activity_main)
        BroadcastDispatcher.registerScreenOff(this)
        registerNetChange()

        initView()
    }

    private fun registerNetChange() {
        val intentFilter = IntentFilter()
        intentFilter.addAction("android.net.conn.CONNECTIVITY_CHANGE")
        connectionChangeReceiver = ConnectionChangeReceiver()
        registerReceiver(connectionChangeReceiver, intentFilter)
    }

    private fun initView() {
        mPlayer = MediaPlayer()
        val progressBar = ProgressBar(this).apply {
            isIndeterminate = true
        }
        progressDialog = AlertDialog.Builder(this)
            .setTitle("云端检测中")
            .setView(progressBar)
            .setCancelable(false)
            .create()

        mSpaceLivenessView = findViewById(R.id.space_liveness_view)
        mFaceDetectView = findViewById(R.id.face_detect_view)
        imgBtnBack = findViewById(R.id.img_btn_back)
        ivVoice = findViewById(R.id.iv_voice)
        surfaceView = findViewById(R.id.surface_view)
        pvCountTime = findViewById(R.id.pv_count_time)
        gifAction = findViewById(R.id.gif_action)
        detectAura = findViewById(R.id.detect_aura)
        tvTip = findViewById(R.id.tv_tip)
        imgBtnBack?.setOnClickListener {
            mAliveDetector?.stopDetect()
            finish()
        }

        ivVoice?.setOnClickListener {
            isOpenVoice = !isOpenVoice
            if (isOpenVoice) {
                ivVoice?.setImageResource(R.mipmap.ico_voice_open_2x)
            } else {
                ivVoice?.setImageResource(R.mipmap.ico_voice_close_2x)
            }
        }

        BroadcastDispatcher.addScreenStatusChangedListener(object :
            BroadcastDispatcher.ScreenStatusChangedListener {
            override fun onForeground() {
                resetIndicator()
                resetGif()
                mAliveDetector?.startDetect()
            }

            override fun onBackground() {
                mAliveDetector?.stopDetect()
            }

        })

        initData()
    }

    private fun initData() {
        mAliveDetector = AliveDetector.getInstance()
        mAliveDetector?.setDebugMode(true)
        mAliveDetector?.init(this, surfaceView, "易盾业务id")
        mAliveDetector?.setDetectedListener(object : DetectedListener {
            override fun onReady(isInitSuccess: Boolean) {
                // 开始倒计时
                pvCountTime?.startCountTimeAnimation()
                // 引擎初始化完成
                if (isInitSuccess) Log.d(TAG, "活体检测引擎初始化完成") else Log.e(
                    TAG,
                    "活体检测引擎初始化失败"
                )
            }

            override fun onActionCommands(actionTypes: Array<ActionType>?) {
                // 此次活体检测下发的待检测动作指令序列
                mActions = actionTypes
                val commands = buildActionCommand(actionTypes)
                Log.d(TAG, "活体检测动作序列为:$commands")
            }

            override fun onStateTipChanged(actionType: ActionType, stateTip: String?, code: Int) {
                // 单步动作
                Log.d(TAG, "actionType:" + actionType.actionTip + " stateTip:" + stateTip)

                dealWithTipChanged(actionType, stateTip)
            }

            override fun onPassed(isPassed: Boolean, token: String?) {
                // 检测通过
                dealWithPassed(isPassed, token)
            }

            override fun onCheck() {
                // 云端检测
                if (!isFinishing) {
                    progressDialog?.show()
                }
            }

            override fun onError(code: Int, msg: String?, token: String?) {
                if (progressDialog?.isShowing == true) {
                    progressDialog?.dismiss()
                }
                Log.e(TAG, "listener [onError] 活体检测出错,原因:$msg token:$token")
            }

            override fun onOverTime() {
                Util.showDialog(
                    this@MainActivity, "检测超时", "请在规定时间内完成动作",
                    "重试", "返回首页", { _, _ ->
                        resetIndicator()
                        resetGif()
                        mAliveDetector?.startDetect()
                    }) { _, _ ->
                    val intent = Intent(
                        this@MainActivity,
                        WelcomeActivity::class.java
                    )
                    startActivity(intent)
                }
            }

            override fun onBackgroundColor(color: Int) {
                // -1说明是动作活体切换
                if (color != -1) {
                    Glide.with(this@MainActivity).load(R.mipmap.pic_front_2x).into(gifAction!!)
                    tvTip?.text = ""
                    detectAura?.visibility = View.VISIBLE
                    detectAura?.start(color)
                } else {
                    detectAura?.let {
                        if (it.visibility == View.VISIBLE) {
                            it.visibility = View.GONE
                        }
                    }
                }
            }

            override fun onSpaceLiveness(direction: Int) {
                enterSpaceModelUI()
            }

            override fun onSpaceStateTipChanged(
                actionType: ActionType?,
                stateTip: String?,
                code: Int
            ) {
                if (actionType == ActionType.ACTION_PASSED) {
                    exitSpaceModelUI()
                }
                when (actionType) {
                    ActionType.ACTION_SPACE_NEAR, ActionType.ACTION_SPACE_FAR -> {
                        when (code) {
                            100 -> {
                                mSpaceLivenessView?.reset()
                                mLastSpaceCode = -1
                            }

                            7 -> {
                                mSpaceLivenessView?.setDistanceType(SpaceLivenessView.DISTANCE_NEAR)
                                if (mLastSpaceCode != 7) mSpaceLivenessView?.startNearAnimation()
                                mSpaceLivenessView?.setRingType(SpaceLivenessView.RING_TYPE_COLOR)
                                mSpaceLivenessView?.setTipSecondText(stateTip ?: "请略微靠近屏幕")
                                mLastSpaceCode = 7
                            }

                            8 -> {
                                mSpaceLivenessView?.setDistanceType(SpaceLivenessView.DISTANCE_FAR)
                                if (mLastSpaceCode != 8) mSpaceLivenessView?.startFarAnimation()
                                mSpaceLivenessView?.setRingType(SpaceLivenessView.RING_TYPE_COLOR)
                                mSpaceLivenessView?.setTipSecondText(stateTip ?: "请略微远离屏幕")
                                mLastSpaceCode = 8
                            }

                            0 -> {
                                mSpaceLivenessView?.setRingType(SpaceLivenessView.RING_TYPE_BLUE)
                                mSpaceLivenessView?.setTipSecondText(stateTip ?: "请保持不动")
                                mLastSpaceCode = 0
                            }

                            else -> {}
                        }
                    }

                    else -> {}
                }
            }
        })

        mAliveDetector?.setTimeOut(30000)
        mAliveDetector?.startDetect()
    }

    /**
     * 处理单步动作
     */
    private fun dealWithTipChanged(actionType: ActionType, stateTip: String?) {
        when (actionType) {
            ActionType.ACTION_ERROR -> setTipText(stateTip, true)
            ActionType.ACTION_PASSED -> {
                Log.d(TAG, "检测通过")
            }

            else -> setTipText(stateTip, false)
        }

        if (actionType != ActionType.ACTION_PASSED && actionType != ActionType.ACTION_ERROR) {
            if (actionType != this.mCurrentActionType) {
                mCurrentActionType = actionType
                when (actionType) {
                    ActionType.ACTION_TURN_HEAD_TO_LEFT -> {
                        gifAction?.let {
                            Glide.with(applicationContext).asGif().load(R.drawable.turn_left)
                                .into(gifAction!!)
                        }
                        if (isOpenVoice) {
                            playSound(getAssetFileDescriptor("turn_head_to_left.wav"))
                        }
                    }

                    ActionType.ACTION_TURN_HEAD_TO_RIGHT -> {
                        gifAction?.let {
                            Glide.with(applicationContext).asGif().load(R.drawable.turn_right)
                                .into(gifAction!!)
                        }
                        if (isOpenVoice) {
                            playSound(getAssetFileDescriptor("turn_head_to_right.wav"))
                        }
                    }

                    ActionType.ACTION_OPEN_MOUTH -> {
                        gifAction?.let {
                            Glide.with(applicationContext).asGif().load(R.drawable.open_mouth)
                                .into(gifAction!!)
                        }
                        if (isOpenVoice) {
                            playSound(getAssetFileDescriptor("open_mouth.wav"))
                        }
                    }

                    ActionType.ACTION_BLINK_EYES -> {
                        gifAction?.let {
                            Glide.with(applicationContext).asGif().load(R.drawable.open_eyes)
                                .into(gifAction!!)
                        }
                        if (isOpenVoice) {
                            playSound(getAssetFileDescriptor("blink_eyes.wav"))
                        }
                    }

                    else -> {
                        Log.d(TAG, "不支持的类型")
                    }
                }
            }
        }
    }

    /**
     * 处理本地检测通过
     */
    private fun dealWithPassed(isPassed: Boolean, token: String?) {
        // 检测通过
        if (progressDialog?.isShowing == true) {
            progressDialog?.dismiss()
        }
        if (isPassed) {
            Log.d(TAG, "活体检测通过,token is:$token")
            finish()
            val intent = Intent(this@MainActivity, SuccessActivity::class.java)
            startActivity(intent)
        } else {
            Log.e(TAG, "活体检测不通过,token is:$token")
            finish()
            val intent = Intent(
                this@MainActivity,
                FailureActivity::class.java
            )
            intent.putExtra("token", token)
            startActivity(intent)
        }
    }

    /**
     * actionIds
     */
    private fun buildActionCommand(actionCommands: Array<ActionType>?): String {
        val commands = StringBuilder()
        actionCommands?.let {
            for (actionType in it) {
                commands.append(actionType.actionID)
            }
        }
        return if (TextUtils.isEmpty(commands.toString())) "" else commands.toString()
    }

    private fun resetIndicator() {
        mCurrentActionType = ActionType.ACTION_STRAIGHT_AHEAD
        mSpaceLivenessView?.visibility = View.GONE
        mSpaceLivenessView?.reset()
        mFaceDetectView?.visibility = View.VISIBLE
        mFaceDetectView?.reset()
    }

    private fun enterSpaceModelUI() {
        runOnUiThread {
            mLastSpaceCode = -1
            mSpaceLivenessView?.visibility = View.VISIBLE
            mSpaceLivenessView?.reset()
            mFaceDetectView?.visibility = View.GONE
            mFaceDetectView?.reset()
        }
    }

    private fun exitSpaceModelUI() {
        runOnUiThread {
            mSpaceLivenessView?.visibility = View.GONE
            mSpaceLivenessView?.reset()
            mFaceDetectView?.visibility = View.VISIBLE
        }
    }

    private fun resetGif() {
        gifAction?.let {
            Glide.with(applicationContext).load(R.mipmap.pic_front_2x).into(it)
        }
    }

    private fun playSound(fileDescriptor: AssetFileDescriptor?) {
        try {
            mPlayer?.reset()
            fileDescriptor?.let {
                mPlayer?.setDataSource(
                    it.fileDescriptor,
                    it.startOffset,
                    it.length
                )
            }
            mPlayer?.prepare()
            mPlayer?.start()
        } catch (e: IOException) {
            e.printStackTrace()
            Log.e(TAG, "playSound error$e")
        }
    }

    private fun getAssetFileDescriptor(assetName: String): AssetFileDescriptor? {
        try {
            return application.assets.openFd(assetName)
        } catch (e: IOException) {
            e.printStackTrace()
            Log.e(TAG, "getAssetFileDescriptor error$e")
        }
        return null
    }

    @SuppressLint("SetTextI18n")
    private fun setTipText(tip: String?, isErrorType: Boolean) {
        if (isErrorType) {
            val display = when (tip) {
                "请移动人脸到摄像头视野中间" -> "请正对手机屏幕\n将面部移入框内"
                "请正视摄像头视野中间并保持不动" -> "请正视摄像头\n并保持不动"
                else -> tip ?: ""
            }
            mFaceDetectView?.setTipTopText(display)
            mSpaceLivenessView?.setTipTopText(display)
        } else {
            mFaceDetectView?.setTipTopText("")
            mFaceDetectView?.setTipSecondText(tip ?: "")
            mSpaceLivenessView?.setTipTopText("")
        }
    }

    override fun onPause() {
        super.onPause()

        progressDialog?.let {
            if (it.isShowing) {
                it.dismiss()
            }
        }
        if (mPlayer?.isPlaying == true) {
            mPlayer?.pause()
        }
    }

    override fun onStop() {
        super.onStop()
        mPlayer?.stop()
    }

    override fun onDestroy() {
        Util.setWindowBrightness(this, WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE)
        if (isFinishing) {
            mAliveDetector?.stopDetect()
            mAliveDetector?.destroy()
            mAliveDetector = null
            pvCountTime?.cancelCountTimeAnimation()
            BroadcastDispatcher.unRegisterScreenOff(this)
            connectionChangeReceiver?.let {
                unregisterReceiver(it)
            }
        }

        if (progressDialog?.isShowing == true) {
            progressDialog?.dismiss()
        }
        progressDialog = null
        if (mPlayer?.isPlaying == true) {
            mPlayer?.stop()
        }
        mPlayer?.reset()
        mPlayer?.release()
        mPlayer = null
        super.onDestroy()
    }
}
